Add a "view as BibTeX" option before importing an entry from the cita…

JamseBonde007 opened the pull request:
Add a "view as BibTeX" option before importing an entry from the cita…tion relation tab  #11826

This PR implements the feature requested in issue #10869. It adds a button that allows users to display a dialog showing the full BibTeX entry when importing citations from the Citation Relation tab. This enhancement will help users decide whether to import a potentially "duplicate" record by allowing them to see the full details of the entry in BibTeX format.

![image](https://github.com/user-attachments/assets/6e106c72-bdb2-4e91-aee5-6247e46ebe03)
![image](https://github.com/user-attachments/assets/72b7fadd-c196-4e80-8032-f708abb9cc3e)

Closes #11826


### Mandatory checks

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

koppor commented on src/main/resources/l10n/JabRef_en.properties:
```diff
@@ -2782,3 +2782,8 @@ Warning\:\ The\ selected\ directory\ is\ not\ a\ valid\ directory.=Warning: The
 Currently\ selected\ CSL\ Style\:\ '%0' = Currently selected CSL Style: '%0'
 Store\ url\ for\ downloaded\ file=Store url for downloaded file
+
+BibTeX\ of\ that\ entry=BibTeX of that entry
```
Re-use "Show BibTeX source" (or "Show %0 source". Just check the existing strings.)

JamseBonde007 replied:
Exits only "Show/edit %0 source"

koppor commented on src/main/resources/l10n/JabRef_en.properties:
```diff
@@ -2782,3 +2782,8 @@ Warning\:\ The\ selected\ directory\ is\ not\ a\ valid\ directory.=Warning: The
+
+BibTeX\ of\ that\ entry=BibTeX of that entry
+Could\ not\ load\ %0\ source=Could not load %0 source
+BibTeX\ source=BibTeX source
```
Please check the existing one. I think, it is `%0 source`.

koppor commented on src/main/resources/l10n/JabRef_en.properties:
```diff
@@ -2782,3 +2782,8 @@ Warning\:\ The\ selected\ directory\ is\ not\ a\ valid\ directory.=Warning: The
 Store\ url\ for\ downloaded\ file=Store url for downloaded file
+
+BibTeX\ of\ that\ entry=BibTeX of that entry
+Could\ not\ load\ %0\ source=Could not load %0 source
```
Not more information than "Error". Thus, please use "Error" as string - and not a new text.

koppor commented on src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java:
```diff
@@ -270,6 +293,37 @@ private void styleFetchedListView(CheckListView<CitationRelationItem> listView)
+    private EntryEditor createEntryEditor() {
+        Supplier<LibraryTab> tabSupplier = () -> this.libraryTab;
+        return new EntryEditor(this.libraryTab,
+                // Actions are recreated here since this avoids passing more parameters and the amount of additional memory consumption is neglegtable.
```
I think, the main reason is that undo and redo are independent here. Thus, these should be fake ones? -- The BibTeX is from the external service and NOT contained in the current library. Thus, it has NOTHING to do with the current library and thus no undo/redo relation with it.

**Update**: I think, this is not necessary at all. No entry editor is needed.

koppor commented on src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java:
```diff
@@ -270,6 +293,37 @@ private void styleFetchedListView(CheckListView<CitationRelationItem> listView)
+    }
+
+    private void showEntrySourceDialog(CodeArea codeArea) {
+        if (codeArea == null) {
+            dialogService.showWarningDialogAndWait(Localization.lang("BibTeX source", databaseContext.getMode().getFormattedName()), Localization.lang("Could not load %0 source", databaseContext.getMode().getFormattedName()));
```
How often does this occur? Bug in the code or real use? I think, its bug in the code. Thus, please use `LOGGER.warn(...)` instead of a dialog service.

koppor replied:
Also just return then. #FailFast. Do not use `} else { `.

koppor commented on src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java:
```diff
@@ -254,6 +268,15 @@ private void styleFetchedListView(CheckListView<CitationRelationItem> listView)
+                    Button showEntrySource = IconTheme.JabRefIcons.SOURCE.asButton();
+                    showEntrySource.setTooltip(new Tooltip(Localization.lang("Show %0 source", databaseContext.getMode().getFormattedName())));
+                    showEntrySource.setOnMouseClicked(event -> {
+                        this.entryEditor = createEntryEditor();
```
Why do you need a full entry editor? I think, you can just show the source without full editing capabilities.

JamseBonde007 commented:
Thank you for your feedback! I've addressed the issues you pointed out and made the necessary changes. Please take a look and let me know if everything looks good now.

koppor commented on src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java:
```diff
@@ -270,6 +295,40 @@ private void styleFetchedListView(CheckListView<CitationRelationItem> listView)
+        StringWriter writer = new StringWriter();
+        BibWriter bibWriter = new BibWriter(writer, OS.NEWLINE);
+        FieldWriter fieldWriter = FieldWriter.buildIgnoreHashes(this.preferences.getFieldPreferences());
+        new BibEntryWriter(fieldWriter, new BibEntryTypesManager()).write(entry, bibWriter, type);
```
Can you try to get the `BibEntryTypesManager` be passed? It should be available "somewhere" in the call hierarchy. `org.jabref.gui.collab.entrychange.PreviewWithSourceTab` also has it.

koppor replied:
Please add an `@implNote` that this code is similar to `PreviewWithSourceTab#getSourceString`. 

@Siedlerchr I think it is OK to have code duplication here. Alternatively, we could introduce `org.jabref.logic.exporter.BibEntryPrinter` having a single method `getAsString`.

Siedlerchr replied:
Duplicated code is imho okay, just the EntryTypesManaager as constructor parameter

koppor commented on src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java:
```diff
@@ -270,6 +295,40 @@ private void styleFetchedListView(CheckListView<CitationRelationItem> listView)
+        DialogPane dialogPane = new DialogPane();
+        dialogPane.setPrefSize(800, 400);
+        dialogPane.setContent(scrollPane);
+        String title = Localization.lang("%0 source", "Show BibTeX");
```
"Show BibTeX" is not localized.

```suggestion
        String title = Localization.lang("Show BibTeX source");
```

koppor commented on src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java:
```diff
@@ -254,6 +271,14 @@ private void styleFetchedListView(CheckListView<CitationRelationItem> listView)
                     }
 
+                    Button showEntrySource = IconTheme.JabRefIcons.SOURCE.asButton();
+                    showEntrySource.setTooltip(new Tooltip(Localization.lang("%0 source", databaseContext.getMode().getFormattedName())));
```
I am not sure if it is really BibLaTeX in this case - the citatoin fetcher always uses the bibtex dialect? Can you check? If unsure, just always use "BibTeX".

koppor commented on src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java:
```diff
@@ -74,6 +87,9 @@ public class CitationRelationsTab extends EntryEditorTab {
     private final DuplicateCheck duplicateCheck;
+    private EntryEditor entryEditor;
+
+    private StateManager stateManager;
```
Not used. Remove.

koppor commented on src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java:
```diff
@@ -74,6 +87,9 @@ public class CitationRelationsTab extends EntryEditorTab {
     private final BibEntryRelationsRepository bibEntryRelationsRepository;
     private final CitationsRelationsTabViewModel citationsRelationsTabViewModel;
     private final DuplicateCheck duplicateCheck;
+    private EntryEditor entryEditor;
```
Not used. Please remove.

It would be really nice if you scrolled through your source code in IntelliJ. That variable should be in light gray meaning: not used. 😅

The thing is that we as reviewers could overlook things. Moreover, if we comment on these kind things, there is no time to check the harder questions... 😇

JamseBonde007 commented:
Thank you for the feedback and the tips! I've addressed the comments and reviewed the code in IntelliJ. Please take a look at the updates and let me know if everything looks good.

Siedlerchr commented on src/main/java/org/jabref/gui/entryeditor/citationrelationtab/CitationRelationsTab.java:
```diff
@@ -103,12 +102,12 @@ public CitationRelationsTab(DialogService dialogService,
         setTooltip(new Tooltip(Localization.lang("Show articles related by citation")));
 
-        this.duplicateCheck = new DuplicateCheck(new BibEntryTypesManager());
+        this.entryTypesManager = new BibEntryTypesManager();
```
You can pass the instance from the caller (see EntryEditor.java)

koppor replied:
Use the instance from EntryEditor as well. Mabe, this will also fix some bugs...

JamseBonde007 commented:
Fixed

Siedlerchr reviewed (commented):
From my point of view looks good!
Thanks for the contribution

Please resolve the comments you addressed, makes it easier for us to see which ones you fixed etc

Siedlerchr commented:
@JamseBonde007  Thanks for your contribution!
